crates/starknet-types-core/src/felt/qm31.rs

```diff
@@ -0,0 +1,487 @@
+use core::fmt;
+
+use crate::felt::Felt;
+
+pub const STWO_PRIME: u64 = (1 << 31) - 1;
+const STWO_PRIME_U128: u128 = STWO_PRIME as u128;
+const MASK_36: u64 = (1 << 36) - 1;
+const MASK_8: u64 = (1 << 8) - 1;
+
+#[derive(Debug)]
+pub enum QM31Error {
+    UnreducedFelt(Felt),
+    FeltTooBig(Felt),
+    InvalidInversion,
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for QM31Error {}
+
+impl fmt::Display for QM31Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            QM31Error::UnreducedFelt(felt) => writeln!(
+                f,
+                "Number is not a packing of a QM31 in reduced form: {felt})"
+            ),
+            QM31Error::FeltTooBig(felt) => writeln!(
+                f,
+                "Number used as QM31 since it's more than 144 bits long: {felt}"
+            ),
+            QM31Error::InvalidInversion => writeln!(f, "Attempt to invert a qm31 equal to zero"),
+        }
+    }
+}
+
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct QM31Felt([u64; 4]);
+
+impl QM31Felt {
+    /// [QM31Felt] constant that's equal to 0.
+    pub const ZERO: Self = Self([0, 0, 0, 0]);
+
+    pub fn as_raw(&self) -> [u64; 4] {
+        self.0
+    }
+
+    /// Create a [QM31Felt] from the raw internal representation. Reduces four u64 coordinates so that they fit in 144 bits.
+    pub fn from_raw(coordinates: [u64; 4]) -> QM31Felt {
+        Self([
+            coordinates[0] % STWO_PRIME,
+            coordinates[1] % STWO_PRIME,
+            coordinates[2] % STWO_PRIME,
+            coordinates[3] % STWO_PRIME,
```

Comment:
This is a one-way destructive operation.
Naming it `from_raw`, given that there is an `as_raw` method, imply that 
`x == QM31Felt::from_raw(x.as_raw())`
Which is not always true. 
Should probably rename it `from_coordinates`